x-pack/filebeat/module/aws/_meta/docs.asciidoc

```diff
@@ -128,6 +146,24 @@ Default to be 300 seconds.
 
 Maximum duration before AWS API request will be interrupted. Default to be 120 seconds.
 
+*`var.s3_bucket`*::
+
+AWS S3 bucket ARN (Required when `var.queue_url` is not set).
+
+*`var.s3_bucket_poll_interval`*::
+
+Interval between list requests to the S3 bucket. Default to be 120 seconds.
+
+*`var.s3_bucket_objects_expiration`*::
+
+Expiration of the S3 objects in the bucket. If you have an expiration set on the
+S3 objects in the bucket you can set it here so that internal status of the object will
```

Comment:
I didn't know this exists, nice. Do you have any docs on the AWS side we can reference? Also, do you know how common it is to use this as compared to just "keep everything" in the bucket? I guess this has some implications on our thinking around the state store.

I'm mostly concerned about forever growing state in high traffic scenarios

Reply:
https://aws.amazon.com/it/blogs/aws/amazon-s3-object-expiration/

I will add the link to docs as well

I'm not sure about how much is common indeed. The problem with not expiring objects, once they are ingested, if we set a TTL in order to flush from the state store they would be ingested again.
It is a trade off between repeated ingestions and having a state store that keeps growing (I wrote in the documentation that this has impact on memory usage of filebeat).
If ingestion of the same data is idempotent, we could define this TTL as "forget status" for the s3 object, explain better the implications and how the best setup is to link it to the s3 object expiration.
But what if the data were purged from elastic? After a little time they would reappear and this can be confusing for the users

Reply:
how about doing keeping a time based cursor after each run? Once a full object list / retrieval is completed we could potentially remove all the states and replace them with a single timestamp mark meaning: anything older than this has been shipped already

Reply:
State is checked on listing: https://github.com/elastic/beats/blob/db7e44cfd921cc8f853f18d60305d8df6ae3365d/x-pack/filebeat/input/awss3/collector.go#L140-L142

If I got it right, we should add here that even if it is not new (in this case because the state was purge because stored) but `state.LastModified` is before the last listing we should skip anyway because it is shipped already.
We should have a specific check on state to ensure that the previous state we compare doesn't exists.
Yes, I think it would be a great improvement.

Reply:
> we should add here that even if it is not new

I think you actually mean "new". but yes, I believe we are in agreement.

To me this approach raises the question: should he have the TTL option at all? or just rely on doing cleanup + keeping this `lastModified` cursor after each complete fetch?